core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java

```diff
@@ -214,6 +216,15 @@ public void handleException(TransportException exp) {
         void retry(final @Nullable Throwable failure) {
             if (observer.isTimedOut()) {
                 // we running as a last attempt after a timeout has happened. don't retry
+                Throwable listenFailure = failure;
+                if (listenFailure == null) {
+                    if (shardIt == null) {
+                        listenFailure = new ElasticsearchTimeoutException("Timeout waiting for [" + request.timeout() + "], request: " + request.toString());
+                    } else {
+                        listenFailure = new ElasticsearchTimeoutException("[" + shardIt.size() + "] shardIt, [" + shardIt.sizeActive() + "] active : Timeout waiting for [" + request.timeout() + "], request: " + request.toString());
+                    }
+                }
+                listener.onFailure(listenFailure);
```

Comment:
also, can we remove the boolean return value from doStart and remove the timeout handling from the public void onTimeout(TimeValue timeout)  method of the callback given to the observer in line 245? just call doStart.